libp2p/protocols/connectivity/autonatv2/client.nim

```diff
@@ -0,0 +1,165 @@
+# those terms.
+
+{.push raises: [].}
+
+import results
+import chronos, chronicles
+import
+  ../../protocol,
+  ../../../switch,
+  ../../../multiaddress,
+  ../../../multicodec,
+  ../../../peerid,
+  ../../../protobuf/minprotobuf,
+  ./types
+
+logScope:
+  topics = "libp2p autonat v2 client"
+
+const
+  MaxAcceptedDialDataRequest* = 100 * 1024 # 100 KB
+  MaxDialDataResponsePayload* = 1024
+
+type AutonatV2Client* = ref object of LPProtocol
+  switch*: Switch
+  rng: ref HmacDrbgContext
+  nonces: Table[PeerId, uint64]
+
+proc handleDialBack(
+    self: AutonatV2Client, conn: Connection, dialBack: DialBack
+) {.async: (raises: [CancelledError, AutonatV2Error, LPStreamError]).} =
+  let expectedNonce =
+    try:
+      self.nonces[conn.peerId]
+    except exceptions.KeyError:
+      raise newException(AutonatV2Error, "Not expecting DialBack from " & $conn.peerId)
+
+  self.nonces.del(conn.peerId)
+
+  if expectedNonce != dialBack.nonce:
+    error "Received DialBack with an invalid nonce",
+      receivedNonce = dialBack.nonce, expectedNonce = expectedNonce
+    return
+
+  let resp = DialBackResponse(status: DialBackStatus.Ok)
+  try:
+    trace "Sending DialBackResponse", dialBackResp = resp
+    await conn.writeLp(resp.encode().buffer)
+  except CancelledError as exc:
+    raise exc
+  except CatchableError as exc:
+    raise newException(AutonatV2Error, "Sending DialBackResponse failed", exc)
+
+proc new*(T: typedesc[AutonatV2Client], switch: Switch, rng: ref HmacDrbgContext): T =
+  let client = T(switch: switch, rng: rng, nonces: initTable[PeerId, uint64]())
+
+  # handler for DialBack messages
+  proc handleStream(
+      conn: Connection, proto: string
+  ) {.async: (raises: [CancelledError]).} =
+    try:
+      let dialBack = DialBack.decode(initProtoBuffer(await conn.readLp(1024))).valueOr:
+        raise newException(AutonatV2Error, "Unable to decode DialBack")
+      debug "Received DialBack", nonce = $dialBack.nonce
+      await client.handleDialBack(conn, dialBack)
+    except CancelledError as exc:
+      trace "handler cancelled"
+      raise exc
+    except CatchableError as exc:
+      debug "exception in handler", description = exc.msg, conn
+    except LPStreamRemoteClosedError as exc:
+      error "connection closed by peer", description = exc.msg, peer = conn.peerId
+    finally:
+      trace "closing connection", conn
+      await conn.close()
+
+  client.handler = handleStream
+  client.codec = $AutonatV2Codec.DialBack
+  client
+
+proc handleDialDataRequest*(
+    conn: Connection, req: DialDataRequest
+): Future[DialResponse] {.
+    async: (raises: [CancelledError, AutonatV2Error, LPStreamError])
+.} =
+  debug "Received DialDataRequest",
+    numBytes = req.numBytes, maxAcceptedNumBytes = MaxAcceptedDialDataRequest
+
+  if req.numBytes > MaxAcceptedDialDataRequest:
+    raise newException(
+      AutonatV2Error, "Rejecting DialDataRequest: numBytes is greater than the maximum"
+    )
+
```

Comment:
We should probably validate too that the AddrIndex received here is valid (maybe it'a negative number or a number greater than the number of addresses we sent)